proposals/extensions.md

```diff
@@ -399,54 +399,194 @@ These requirements need more refinement as implementation progresses, and may ne
 
-#### Extension types
+Here's a schematized overview of metadata encoding:
+```
+[Extension]
+static class EnclosingStaticClass
+{
+    [Extension]
+    public class ExtensionGroupingType1 // has type parameters with minimal constraints sufficient to keep extension member declarations below valid
+    {
+        private static class ExtensionMarkerType1 // has re-declared type parameters with full fidelity of C# constraints
+        {
+            private static void <Extension>$(... extension parameter ...) // extension marker method
+        }
+        ... ExtensionMarkerType2, etc ...
+
+        ... extension members for ExtensionGroupingType1, each points to its corresponding extension marker type ...
+    }
+
+    ... ExtensionGroupingType2, etc ...
+
+    ... implementation methods ...
+}
+```
+
+The enclosing static class is emitted with an `[Extension]` attribute.  
+
+Note: This allows roundtripping of extension declaration symbols through metadata (full and reference assemblies).  
+
+### CLR-level signature vs. C#-level signature
+
+The CLR-level signature of an extension block results from:
+- normalizing type parameter names to `T0`, `T1`, etc ...
+- removing attributes
+- erasing the parameter name
+- erasing parameter modifiers (like `ref`, `in`, `scoped`, ...)
+- erasing tuple names
+- erasing nullability annotations
+- erasing `notnull` constraints
+
+Note: other constraints are preserved, such as `new()`, `struct`, `class`, `allows ref struct`, `unmanaged`, and type constraints.
 
-Each extension declaration in source is emitted as an extension declaration in metadata (sometimes referred to as skeleton type).  
-- Its name is unspeakable and determined based on the lexical order in the program.  
-  The name is not guaranteed to remain stable across re-compilation. 
-  Below we use `<>E__` followed by an index. For example: `<>E__2`.  
-- Its type parameters are those declared in source (including attributes).  
+#### Extension grouping types
+
+Each extension block with a distinct CLR-level signature in source is emitted as an extension grouping type in metadata (extension grouping types are an evolution of the previous "skeleton type" design).  
+- Its name is unspeakable and determined based on the contents of the CLR-level signature. More details below.  
+- Its type parameters have normalized names (`T0`, `T1`, ...) and have no attributes.  
 - Its accessibility is public.
+- It is marked with the `specialname` flag and an `[Extension]` attribute.  
+
+The content-based name of the extension grouping type is based on the CLR-level signature and includes the following:
+- The fully qualified CLR name of the type. This will include namespaces, type constraints, named constraints (like `new`).
```

Comment:
>This will include namespaces, type constraints, named constraints (like `new`). [](http://example.com/codeflow?start=44&length=79)

I am not sure what this means. Constraints for type parameters have their own bullet point below. Why would we want to include constraints declared by extended type (this is different from constraints declared by an extension block)?